File: Python/2_AddTwoNumbers.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: Optional[ListNode]
        :type l2: Optional[ListNode]
        :rtype: Optional[ListNode]
        """

        sul = ListNode(0) #sum list 
        head = sul

        cv = 0 # a value to carry over

        while l1 != None or l2 != None or cv != 0:
            su = cv
            cv = 0

            if l1 != None:
                su += l1.val
            if l2 != None:
                su += l2.val
            
            if su < 10:
                head.val = su
            else:
                head.val = su - 10
                cv = 1

            if l1 != None:
                l1 = l1.next
            if l2 != None:
                l2 = l2.next

            if l1 != None or l2 != None or cv != 0:    
                head.next = ListNode(0)
                head = head.next

        return sul
        
# algorithm: make new list carring over a value if an added result is over 10
# confirm the cases, such as inconsistent length, carrying over 1

File: C++/2_AddTwoNumbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:

    int countNode(ListNode* l) {
        int num = 0;

        while (l != nullptr) {
            num++;
            l = l->next;
        }
        return num;
    }

    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int c = 0;  // carry over
        int val = 0;

        ListNode* head;
        ListNode* main;

        if (countNode(l1) < countNode(l2)) main = l2;
        else main = l1;

        head = main;

        while(l1 != nullptr || l2 != nullptr) {
            
            val = c;
            c = 0;

            if (l1 != nullptr) val += l1->val;
            if (l2 != nullptr) val += l2->val;
            if (val > 9) {
                val -= 10;
                c = 1;
            }

            main->val = val;

            if (l1 != nullptr) l1 = l1->next;
            if (l2 != nullptr) l2 = l2->next;

            if (l1 == nullptr && l2 == nullptr && c != 0) {
                ListNode* newNode = new ListNode(c);
                c = 0;
                main->next = newNode;
            }
            main = main->next;
        }

        return head;
    }
};

// 2,4,9
// 5,6
// 7,0,0,1